api/src/user/guards/ability.guard.ts

```diff
@@ -40,17 +48,43 @@ export class Ability implements CanActivate {
       return true;
     }
 
-    const { user, method, _parsedUrl, session } = context
+    const { user, method, _parsedUrl, session, query } = context
       .switchToHttp()
       .getRequest<Request & { user: User; _parsedUrl: Url }>();
 
-    if (!user) {
+    const paths = _parsedUrl.pathname.split('/');
+    const modelFromPathname = paths?.[1].toLowerCase() as TModel;
+    const isAttachmentUrl =
+      modelFromPathname === 'attachment' &&
+      ['upload', 'download'].includes(paths?.[2]);
+
+    if (!user && !isAttachmentUrl) {
       throw new UnauthorizedException();
     }
     if (!session?.cookie || session.cookie.expires < new Date()) {
       throw new UnauthorizedException('Session expired');
     }
 
+    if (isAttachmentUrl) {
+      // attachment
+      const attachmentUploadContext =
+        query?.context?.toString() as TContextType;
+      if (
+        method === 'POST' &&
+        paths?.[2] === 'upload' &&
+        attachmentUploadContext
+      )
+        return await this.hasRequiredUploadPermission(
+          user,
+          attachmentUploadContext,
+        );
+      else if (method === 'GET' && paths?.[2] === 'download')
+        return await this.hasRequiredDownloadPermission(user, paths?.[3]);
+      else {
+        return false;
+      }
+    }
+
```

Comment:
Let's move to the parent class in a protected method called "super.canActivate(...)" or a seperate service or class.

So that here you just call it.